### PR TITLE
feat: add sa token component to kfp-persistence

### LINKED ISSUE #343: kfp-persistence stuck in waiting status
### Bug Description

The changes introduced by #331 are affecting the charm, preventing the service to start correctly (replan).

### To Reproduce

1. git checkout kf-3886-release-1.8-update-charms
2. tox -e kfp-persistence-integration -- --keep-models # keep-models for debugging purposes
3. wait for everything to settle
4. `kfp-persistence` will be stuck in waiting status with message:

```
[container:persistenceagent] Waiting for Pebble services (persistenceagent).  If this persists, it could be a blockin...
```

### Environment

Ubuntu 20.04
juju 3.1
microk8s 1.25-strict/stable

### Relevant log output

```shell
unit-kfp-persistence-0: 10:33:42 ERROR unit.kfp-persistence/0.juju-log kfp-api:5: execute_components caught unhandled exception when executing configure_charm for container:persistenceagent
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-kfp-persistence-0/charm/venv/charmed_kubeflow_chisme/components/charm_reconciler.py", line 92, in reconcile
    component_item.component.configure_charm(event)
  File "/var/lib/juju/agents/unit-kfp-persistence-0/charm/venv/charmed_kubeflow_chisme/components/component.py", line 50, in configure_charm
    self._configure_unit(event)
  File "/var/lib/juju/agents/unit-kfp-persistence-0/charm/venv/charmed_kubeflow_chisme/components/pebble_component.py", line 131, in _configure_unit
    self._update_layer()
  File "/var/lib/juju/agents/unit-kfp-persistence-0/charm/venv/charmed_kubeflow_chisme/components/pebble_component.py", line 142, in _update_layer
    container.replan()
  File "/var/lib/juju/agents/unit-kfp-persistence-0/charm/venv/ops/model.py", line 1915, in replan
    self._pebble.replan_services()
  File "/var/lib/juju/agents/unit-kfp-persistence-0/charm/venv/ops/pebble.py", line 1680, in replan_services
    return self._services_action('replan', [], timeout, delay)
  File "/var/lib/juju/agents/unit-kfp-persistence-0/charm/venv/ops/pebble.py", line 1761, in _services_action
    raise ChangeError(change.err, change)
ops.pebble.ChangeError: cannot perform the following tasks:
- Start service "persistenceagent" (cannot start service: exited quickly with code 1)
----- Logs from task 0 -----
2023-10-02T10:33:42Z INFO Most recent service output:
    W1002 10:33:42.193357      16 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
    time="2023-10-02T10:33:42Z" level=error msg="Error reading persistence agent service account token '/var/run/secrets/kubeflow/tokens/persistenceagent-sa-token': open /var/run/secrets/kubeflow/tokens/persistenceagent-sa-token: no such file or directory"
    time="2023-10-02T10:33:42Z" level=fatal msg="Error starting Service Account Token Refresh Ticker due to: open /var/run/secrets/kubeflow/tokens/persistenceagent-sa-token: no such file or directory"
2023-10-02T10:33:42Z ERROR cannot start service: exited quickly with code 1
-----
```

### Additional context

For more logs and relevant outputs please check the failed CI on https://github.com/canonical/kfp-operators/pull/331

### PR DESCRIPTION
**Pull request by DnPlas**
The kfp-persistence charms needs a ServiceAccount token to be inside the workload container to be able to start the service. The sa token component will generate the required token and save it in a file where it can be accessed to later be pushed from the charm to the workload container.

#### Testing

1. Build and deploy the charm in this PR
2. Using `kubectl exec` check the files inside the workload container. Make sure you check the `persistenceagent` container and not the `charm`.
3. There must be a non empty token file in `/var/run/secrets/kubeflow/tokens/<token name>`

Fixes #343

### CODE REVIEW

File: charms/kfp-persistence/src/templates/auth_manifests.yaml.j2 [hunk 1]
```diff
@@ -54,5 +54,7 @@ subjects:
 apiVersion: v1
 kind: ServiceAccount
 metadata:
-  name: {{ app_name }}-sa
+  labels:
+    app: {{ app_name }}
+  name: {{ sa_name }}
```
Comment (orfeas-k) [inline charms/kfp-persistence/src/templates/auth_manifests.yaml.j2:59]: just a question, why is this needed?

Comment (DnPlas) [reply charms/kfp-persistence/src/templates/auth_manifests.yaml.j2:59]: What part, the whole ServiceAccount?

Comment (orfeas-k) [reply charms/kfp-persistence/src/templates/auth_manifests.yaml.j2:59]: the removal of `-sa` from the service account name

Comment (DnPlas) [reply charms/kfp-persistence/src/templates/auth_manifests.yaml.j2:59]: Oh, well that is because I'm keeping the ServiceAccount name the same as in upstream, see [here](https://github.com/canonical/kfp-operators/pull/349/files#diff-30a280bd44c1418d68f3c82b68265bc3bffec6bf25d23c23e342ab71e09ecf38R35) and [here](https://github.com/kubeflow/manifests/blob/master/apps/pipeline/upstream/base/pipeline/ml-pipeline-persistenceagent-sa.yaml#L4)

File: charms/kfp-persistence/src/components/sa_token_component.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            logger.info("Token file already exists, nothing else to do.")
+        api_response = self._create_sa_token()
+        token = api_response.status.token
+        with open(Path(path, filename), "w") as token_file:
+            token_file.write(token)
+
+    def get_status(self) -> StatusBase:
+        """Return the status of this component."""
+        try:
+            self._generate_and_save_token(self._path, self._filename)
```
Comment (orfeas-k) [inline charms/kfp-persistence/src/components/sa_token_component.py:102]: Is this how `get_status()` works? I wonder what will happen if `_generate_and_save_token()` raises an error. Shouldn't `_generate_and_save_token()` return a value to update status accordingly? (more of a charm reconciler question)

**Review by orfeas-k — COMMENTED** [review]
So left one more comment which is more of base charm question as to if this should be the way to implement the component. The logic in the rest of the code looks good. 

I will go ahead and deploy charms from the PR in order to verify the functionality. My understanding is that I should:
- look for the file that the component is expected to create
- check if kfp-persistence service has started successfully

Is there something else I should look into?

**Comment by DnPlas** [comment]
> So left one more comment which is more of base charm question as to if this should be the way to implement the component. The logic in the rest of the code looks good.
> 
> I will go ahead and deploy charms from the PR in order to verify the functionality. My understanding is that I should:
> 
>     * look for the file that the component is expected to create
> 
>     * check if kfp-persistence service has started successfully
> 
> 
> Is there something else I should look into?

Hey @orfeas-k, I added a "testing" section in the description of the PR. Regarding checking that the service starts, in this case, we'll see the service starting just fine because in pipelines 2.0.0-alpha.7 it does not need this file. We will observe this dependency when we finalise the work in #331. For now just ensuring the file is there is enough.

File: charms/kfp-persistence/src/components/sa_token_component.py (hunk 2 above)
Comment (DnPlas) [reply charms/kfp-persistence/src/components/sa_token_component.py:106]: I did a small refactor to make this easier to read and understand, as well as improve the utilisation of the base charm. I am now generating the sa token file inside the `_configure_app_leader` method in the sa token component (according to the [guide](https://github.com/canonical/charmed-kubeflow-chisme/blob/main/src/charmed_kubeflow_chisme/components/how-to/authoring-a-simple-component.md#defining-what-a-component-does-during-charm-reconciliation)).
Please note I also modified the `get_status()` method so it returns active in the presence of the sa token file, and raises a `GenericCharmRuntime` exception otherwise. About the latter, the base charm currently sets the unit status arbitrarily in the presence of an exception, which we should change in the future.

Comment (orfeas-k) [reply charms/kfp-persistence/src/components/sa_token_component.py:106]: Great addition, thank you @DnPlas. Could we have an issue for this in chisme repo? and we can discuss after release once Andrew is also back.

**Review by orfeas-k — APPROVED** [review]
Great work and collaboration @DnPlas, approving!
